runtime/major_gc.c

```diff
@@ -167,26 +296,56 @@ void caml_darken (value v, value *p /* not used */)
     CAMLassert (!Is_blue_hd (h));
     if (Is_white_hd (h)){
       ephe_list_pure = 0;
+      Hd_val (v) = Blackhd_hd (h);
       if (t < No_scan_tag){
-        Hd_val (v) = Grayhd_hd (h);
-        *gray_vals_cur++ = v;
-        if (gray_vals_cur >= gray_vals_end) realloc_gray_vals ();
-      }else{
-        Hd_val (v) = Blackhd_hd (h);
+        mark_stack_push(Caml_state->mark_stack, v, 0, NULL);
+      }
+    }
+  }
+}
+
+/* This function adds blocks in the passed heap chunk [heap_chunk] to the mark
+   stack. It may return 0 if doing so would cause the mark stack to grow more
+   than a quarter full. This is to lower the chance of triggering another
+   overflow, which would be wasteful. Subsequent calls will continue progress.
+   It will return 1 when the supplied chunk has no more range to redarken. */
```

Comment:
I find your wording rather confusing, it looks like there's a missing case.

```suggestion
/* This function adds blocks in the passed heap chunk [heap_chunk] to
   the mark stack. It returns 1 when the supplied chunk has no more
   range to redarken.  It returns 0 if there are still blocks in the
   chunk that need redarkening because pushing them onto the stack
   would make it grow more than a quarter full. This is to lower the
   chance of triggering another overflow, which would be
   wasteful. Subsequent calls will continue progress.
 */
```